### PR TITLE
Feature: rustg_sound_length()

### PR DESCRIPTION
**Pull request by Kapu1178**
Adds a function for getting the duration of an audio file.
Also adds a function for getting the durations of a list of audio files.

TODO:
- [x] Add a version that accepts a list of filepaths instead.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -123,6 +126,7 @@ http = ["ureq", "serde", "serde_json", "once_cell", "jobs"]
 json = ["serde", "serde_json"]
 log = ["chrono"]
 sanitize = ["ammonia", "serde_json"]
+sound_len = ["symphonia"]
```
Comment (ZeWaka) [inline Cargo.toml:129]: add to readme

File: Cargo.toml [hunk 2]
```diff
@@ -79,6 +80,7 @@ default = [
     "noise",
     "rustls_tls",
     "sanitize",
+    "sound_len",
```
Comment (ZeWaka) [inline Cargo.toml:83]: no

File: dmsrc/soundlen.dm [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/proc/rustg_sound_length(file_path)
```
Comment (ZeWaka) [inline dmsrc/soundlen.dm:1]: documentation

File: src/sound_len.rs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    // Try to decode the data packet
+    let decoded_packet = match decoder.decode(&encoded_packet) {
+        Ok(r) => r,
+        Err(_e) => return Ok(String::from("Failed to decode packet!"))
+    };
+
+    // Grab the sample rate from the spec of the buffer.
+    let sample_rate: f64 = decoded_packet.spec().rate as f64;
+    // Math!
+    let duration_in_seconds: f32 = (samples_capacity / sample_rate) as f32;
```
Comment (ZeWaka) [inline src/sound_len.rs:91]: all of this seems way overkill there has to be a way to do it in the library easily

File: Cargo.toml (hunk 2 above)
Comment (Kapu1178) [reply Cargo.toml:83]: tgstation sound maintainer wants this feature so it makes sense to me

File: src/sound_len.rs (hunk 4 above)
Comment (Kapu1178) [reply src/sound_len.rs]: added a simpler version and kept the codec reader as a fallback

File: dmsrc/sound_len.dm [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+		if(!isfile(file_path))
+			CRASH("rustg_sound_length error: Passed non-text object")
+
+		if(length("[file_path]"))
+			file_path = "[file_path]"
+		else
+			CRASH("rustg_sound_length does not support non-static file refs.")
+
+	if(!isnull((. = sound_cache[file_path])))
+		return .
```
Comment (Mothblocks) [inline dmsrc/sound_len.dm:19]: This code is doing too much and is not easy to understand, please write it out more directly

```suggestion
	var/cached_length = sound_cache[file_path]
	if(!isnull(cached_length))
		return cached_length
```

File: src/sound_len.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    // Gracefully exit if the filepath is invalid.
+    if !path.exists() {
+        return Err(SoundLen(format!("Bad path: {}", sound_path)));
+    }
+
+    // Try to open the file
+    let sound_src = match File::open(path) {
+        Ok(r) => r,
+        Err(_e) => return Err(SoundLen("Couldn't open file!".to_string())),
+    };
```
Comment (Mothblocks) [inline src/sound_len.rs:34]: Nit: It's good practice in general to just open it, and check if the error is file not found. 

https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.NotFound

File: src/sound_len.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Gracefully exit if the filepath is invalid.
+    if !path.exists() {
+        return Err(SoundLen(format!("Bad path: {}", sound_path)));
+    }
+
+    // Try to open the file
+    let sound_src = match File::open(path) {
+        Ok(r) => r,
+        Err(_e) => return Err(SoundLen("Couldn't open file!".to_string())),
```
Comment (Mothblocks) [inline src/sound_len.rs:33]: Why don't we give the error with it? Do we think it'll leak some info?

File: src/sound_len.rs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+fn sound_length_decode(probed: ProbeResult) -> Result<f64> {
+    let mut format = probed.format;
+
+    let track = match format.default_track() {
+        Some(r) => r,
+        None => return Err(SoundLen("Could not get default track".to_string())),
+    };
+
+    // Grab the number of frames of the track
+    let samples_capacity: f64 = if let Some(n_frames) = track.codec_params.n_frames {
```
Comment (Mothblocks) [inline src/sound_len.rs:98]: Shouldn't need `: f64`?

File: src/sound_len.rs (hunk 7 above)
Comment (Kapu1178) [reply src/sound_len.rs:33]: oops, leftover

**Comment by ZephyrTFA** [comment]
holy shit, this will be useful for spatial audio

File: dmsrc/sound_len.dm [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// Provided a static RSC file path or a raw text file path, returns the duration of the file in deciseconds as a float.
+/proc/rustg_sound_length(file_path)
```
Comment (ZephyrTFA) [inline dmsrc/sound_len.dm:2]: can you add support for frequency overrides?

Comment (Kapu1178) [reply dmsrc/sound_len.dm:2]: No.

File: dmsrc/sound_len.dm [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+		sound_cache = list()
+
+	. = 0
+
+	if(!istext(file_path))
+		if(!isfile(file_path))
+			CRASH("rustg_sound_length error: Passed non-text object")
+
+		if(length("[file_path]"))
+			file_path = "[file_path]"
```
Comment (ZeWaka) [inline dmsrc/sound_len.dm:14]: what is this code doing

File: dmsrc/sound_len.dm [hunk 11]
```diff
@@ -0,0 +1,34 @@
+// Provided a static RSC file path or a raw text file path, returns the duration of the file in deciseconds as a float.
```
Comment (ZeWaka) [inline dmsrc/sound_len.dm:1]: code doc
```suggestion
/// Provided a static RSC file path or a raw text file path, returns the duration of the file in deciseconds as a float.
```

File: dmsrc/sound_len.dm [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+	if(isnull(ret))
+		. = 0
+		CRASH("rustg_sound_length error: [ret]")
+
+	sound_cache[file_path] = as_num
+	return as_num
+
+#define RUSTG_SOUNDLEN_SUCCESSES "successes"
+#define RUSTG_SOUNDLEN_ERRORS "errors"
+// Returns a list of lists "successes" and "errors". Successes are file_path : duration. Errors are file_path : error.
```
Comment (ZeWaka) [inline dmsrc/sound_len.dm:33]: code doc and this description is confusing, i don't get the exact data structure

File: src/sound_len.rs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+fn get_sound_length_list(list: &str) -> String {
+    let json: Vec<&str> = match serde_json::from_str(list) {
+        Ok(r) => r,
+        Err(_e) => return String::from("Fatal error: Bad json"),
+    };
+
+    let mut successes: HashMap<String, String> = HashMap::new();
+    let mut errors: HashMap<String, String> = HashMap::new();
+
+    for path_string in json.iter() {
```
Comment (ZeWaka) [inline src/sound_len.rs:138]: probably should be using parallel iter here, giving up free performance

File: src/sound_len.rs [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    Ok(duration_in_desciseconds)
+}
+
+byond_fn!(
+    fn sound_len_list(list) {
+        Some(get_sound_length_list(list))
+    }
+);
+
+fn get_sound_length_list(list: &str) -> String {
```
Comment (ZeWaka) [inline src/sound_len.rs:129]: do you really need all of the explicit types inside this function? pretty sure they can be figured out by the compiler fine

File: src/sound_len.rs [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    // Try to decode the data packet
+    let decoded_packet = match decoder.decode(&encoded_packet) {
+        Ok(r) => r,
+        Err(e) => return Err(SoundLen(format!("Decode error: {}", e))),
+    };
+
+    // Grab the sample rate from the spec of the buffer.
+    let sample_rate = decoded_packet.spec().rate as f64;
+    // Math!
+    let duration_in_desciseconds = samples_capacity / sample_rate * 10.0;
```
Comment (ZeWaka) [inline src/sound_len.rs:119]: parens for math grouping?

**Review by ZeWaka — CHANGES_REQUESTED** [review]

File: dmsrc/sound_len.dm (hunk 10 above)
Comment (Kapu1178) [reply dmsrc/sound_len.dm:14]: This checks if an RSC path is a real file. If it cant be stringified it isn't a file on the disk.

File: src/sound_len.rs (hunk 13 above)
Comment (Kapu1178) [reply src/sound_len.rs:138]: I don't know what this means

File: src/sound_len.rs (hunk 15 above)
Comment (Kapu1178) [reply src/sound_len.rs:119]: Clippy told me the parens were unnecesary

File: dmsrc/sound_len.dm (hunk 10 above)
Comment (ZeWaka) [reply dmsrc/sound_len.dm:14]: then comment that

File: src/sound_len.rs (hunk 13 above)
Comment (ZeWaka) [reply src/sound_len.rs:138]: eh whatever you can always change it if you want it to be faster

**Review by ZeWaka — APPROVED** [review]

**Pull request merged**